Reta agency API

cacraig opened the pull request:
*DO NOT MERGE YET*

This addresses #447 . 

It implements the following endpoints. Using the same endpoint structure as the NIBRS endpoints:

```
/agencies/count/states/suboffenses/<string:state_abbr>/<string:agency_ori>?year=<int:year>
/agencies/count/states/suboffenses/<string:state_abbr>/counties/<string:county_fips_code>?year=<int:year>
/agencies/count/states/offenses/<string:state_abbr>/<string:agency_ori>?year=<int:year>
```

harrisj commented on crime_data/common/newmodels.py:
```diff
@@ -179,6 +179,92 @@ class RetaEstimated(db.Model):
+
+        # Heads up - This is going to probably make local tests fail, as our sample DB's 
+        # only contain a little bit of data - ie. reported may not be 12 (ever).
+        query = query.filter(AgencySums.reported == 12) # Agency reported 12 Months.
```
Yep, this is indeed a problem with local tests. I haven't been able to resolve it yet, but I think I'll have to copy over some data from the full DB

harrisj commented on dba/after_load/create_reta_agency_summary.sql:
```diff
@@ -43,7 +43,7 @@ BEGIN
     SUM(rmos.actual_count) AS actual,
     SUM(rmos.cleared_count) AS cleared,
-    SUM(rmos.juvenile_cleared_count) AS juvenile_cleared
+    SUM(rmos.juvenile_cleared_count) AS juvenile_cleared,
```
Looks like there is a comma to be cleaned up here.

cacraig commented:
@harrisj  - I'm going to add tests, rerun the sql, and update the swagger docs. - I also added lookup by county_fips_code as I am sure that's going to be something we want. Thanks!

cacraig commented:
Holy moly. This got a bit nasty - but the new reworked tables are in prod, and it supports counties as well with reasonable performance. Can you take another look @harrisj  ?

harrisj commented on crime_data/common/newmodels.py:
```diff
@@ -17,7 +17,7 @@
 from crime_data.common.models import RefAgency, RefState, RefCounty
 from crime_data.extensions import db
-from sqlalchemy import or_
+from sqlalchemy import or_,and_
```
Ah, SqlAlchemy... how you vex us

harrisj commented on crime_data/common/newmodels.py:
```diff
@@ -179,6 +179,99 @@ class RetaEstimated(db.Model):
 
     state = relationship(RefState)
 
+class RetaMonthAgencySubcatSummary(db.Model, CreatableModel):
```
I get the need for consistency, but I don't think we need to inherit from CreatableModel here (that was for doing on-the-fly model stuff and technically we don't have detail for agencies at the OffenseSubcat level, so I'm not sure if the name is a bit misleading...

harrisj commented on crime_data/common/newmodels.py:
```diff
@@ -179,6 +179,99 @@ class RetaEstimated(db.Model):
+            query = query.filter(AgencySums.year == year)
+
+        # Heads up - This is going to probably make local tests fail, as our sample DB's 
+        # only contain a little bit of data - ie. reported may not be 12 (ever).
```
Working on fixing this for the future...

harrisj commented on dba/after_load/create_reta_agency_summary.sql:
```diff
@@ -1,18 +1,21 @@
 SELECT rm.data_year,
 rm.agency_id,
-bool_or(CASE WHEN rm.reported_flag = 'Y' THEN TRUE ELSE FALSE END) AS reported
+bool_and(CASE WHEN rm.reported_flag = 'Y' THEN TRUE ELSE FALSE END) AS reported
```
Thank you! Sorry about that. I suppose technically we could also sum up the reportd months and make sure the value == 12 too.

harrisj commented on dba/after_load/create_reta_agency_summary.sql:
```diff
@@ -33,28 +36,103 @@ BEGIN
     RAISE NOTICE 'Executing Inserts for offense_subcat_id: %', i;
-    SET work_mem='2GB';
-    INSERT INTO agency_sums (data_year, agency_id, offense_id, offense_code, reported, unfounded, actual, cleared, juvenile_cleared)  
+    SET work_mem='3GB';
```
The joys of a bigger DB instance

cacraig commented:
Thanks @harrisj  - I implemented your feedback, and added the swagger docs. I think in the mean time this is GTG for our front end folks to play with for a bit. - Still need tests, and such.

harrisj commented:
Okay, I don't want to hold it up from being in production, but please file an issue for tests and you or I can do that in a future PR.
